File: app/Packages/ONTData/Sources/ONTData/Local/DiskRepositories.swift
import Foundation
import ONTKit

/// Le corpus, lu du disque quand il y est, du bundle sinon.
///
/// ## Le bundle n'est pas un repli, c'est le socle
///
/// L'app est livrée avec un corpus complet. Il fait marcher une installation
/// neuve **avant** tout réseau — dans un train, dans un avion, sur un forfait
/// épuisé — et il ne disparaît jamais.
///
/// Ce que [`CorpusUpdater`] télécharge vient le **recouvrir**, fichier par
/// fichier. Un livre présent sur le disque est lu du disque ; les autres
/// continuent de venir du bundle. Il n'y a donc aucun état intermédiaire
/// invalide : à tout instant, chaque livre est lisible dans l'une ou l'autre
/// version, jamais dans aucune.
///
/// ## Pourquoi une réalisation de plus, et pas une option dans l'ancienne
///
/// `CorpusRepository` est un **port**. En écrire une seconde réalisation ne
/// touche ni les vues, ni les modèles, ni les tests des autres : c'est
/// exactement ce que cette architecture permettait, et la première fois qu'on
/// s'en sert.
///
/// `@unchecked Sendable` assumé, comme pour le bundle : le cache est protégé
/// par un verrou, et le contenu décodé est immuable.
public final class DiskCorpusRepository: CorpusRepository, @unchecked Sendable {
    private var dossier: URL
    private let socle: any CorpusRepository
    private let lock = NSLock()
    private var cachedCorpora: [Corpus]?
    private var cachedBooks: [String: Book] = [:]

    public init(
        dossier: URL = CorpusUpdater.dossierParDefaut(),
        socle: any CorpusRepository = BundleCorpusRepository()
    ) {
        self.dossier = dossier
        self.socle = socle
    }

    public func corpora() throws -> [Corpus] {
        lock.lock()
        defer { lock.unlock() }

        if let cachedCorpora { return cachedCorpora }
        let corpora: [Corpus]
        if let file: ONTSchema.CorpusFile = lire("corpus.json") {
            corpora = file.corpora.map(Corpus.init).sorted { $0.order < $1.order }
        } else {
            corpora = try socle.corpora()
        }
        cachedCorpora = corpora
        return corpora
    }

    public func book(_ id: String) throws -> Book {
        lock.lock()
        defer { lock.unlock() }

        if let cached = cachedBooks[id] { return cached }
        let book: Book
        if let dto: ONTSchema.Book = lire("books/\(id).json") {
            book = Book(dto)
        } else {
            book = try socle.book(id)
        }
        cachedBooks[id] = book
        return book
    }

    /// Lit un fichier du disque, ou rend `nil`.
    ///
    /// **Aucune erreur ne remonte**, et c'est délibéré : un fichier absent est
    /// le cas normal — l'app n'a pas encore téléchargé ce livre — et un fichier
    /// illisible ne doit pas empêcher de lire. Dans les deux cas, le socle
    /// répond.
    private func lire<T: Decodable>(_ nom: String) -> T? {
        guard let octets = try? Data(contentsOf: dossier.appendingPathComponent(nom)) else {
            return nil
        }
        return try? JSONDecoder().decode(T.self, from: octets)
    }

    /// Oublie ce qui est en mémoire, après une mise à jour.
    ///
    /// Sans ça, le corpus fraîchement téléchargé n'apparaîtrait qu'au prochain
    /// lancement : les caches tiennent la version d'avant, et rien ne leur dit
    /// qu'elle a vieilli.
    public func oublier() {
        lock.lock()
        defer { lock.unlock() }
        cachedCorpora = nil
        cachedBooks.removeAll()
    }

    /// **Change la source, sans changer le dépôt.**
    ///
    /// Le mode développeur du Mac reconstruit le corpus depuis le vault et
    /// l'écrit à côté du corpus publié. Sans ce point d'entrée, il l'écrivait
    /// et **personne ne le lisait** : le dossier était fixé à la construction.
    ///
    /// Le défaut ne se voyait pas, parce que le bandeau affichait le compte que
    /// le pipeline venait de rendre — un nombre juste, sur un corpus que la
    /// liseuse n'ouvrait pas. C'est l'auteur qui l'a pris, en cherchant le
    /// **chapitre** plutôt que le nombre : « y avait certes écrit 45 mais le
    /// chapitre 20 n'est jamais apparu ».
    ///
    /// On vide les caches dans le même verrou : les rendre séparément
    /// laisserait une fenêtre où le dossier est neuf et le contenu ancien.
    public func regarder(_ nouveau: URL) {
        lock.lock()
        defer { lock.unlock() }
        dossier = nouveau
        cachedCorpora = nil
        cachedBooks.removeAll()
    }
}

/// Le lexique, même principe.
public final class DiskGlossaryRepository: GlossaryRepository, @unchecked Sendable {
    private var dossier: URL
    private let socle: any GlossaryRepository
    private let lock = NSLock()
    private var cachedEntries: [GlossaryEntry]?
    private var cachedOccurrences: [String: [Occurrence]]?

    public init(
        dossier: URL = CorpusUpdater.dossierParDefaut(),
        socle: any GlossaryRepository = BundleGlossaryRepository()
    ) {
        self.dossier = dossier
        self.socle = socle
    }

    public func entries() throws -> [GlossaryEntry] {
        lock.lock()
        defer { lock.unlock() }

        if let cachedEntries { return cachedEntries }
        let entries: [GlossaryEntry]
        if let file: ONTSchema.GlossaryFile = lire("glossary.json") {
            entries = file.entries.map(GlossaryEntry.init)
        } else {
            entries = try socle.entries()
        }
        cachedEntries = entries
        return entries
    }

    public func occurrences(of lemma: String) -> [Occurrence] {
        lock.lock()

        if cachedOccurrences == nil, let file: ONTSchema.OccurrencesFile = lire("occurrences.json") {
            cachedOccurrences = file.byLemma.mapValues { $0.map(Occurrence.init) }
        }
        if let table = cachedOccurrences {
            defer { lock.unlock() }
            return table[lemma] ?? []
        }

        // Rien sur le disque : c'est le socle qui répond, et le verrou est
        // rendu **avant** de l'appeler. Le garder exposerait à un interblocage
        // le jour où le socle voudrait, lui aussi, verrouiller quelque chose.
        lock.unlock()
        return socle.occurrences(of: lemma)
    }

    private func lire<T: Decodable>(_ nom: String) -> T? {
        guard let octets = try? Data(contentsOf: dossier.appendingPathComponent(nom)) else {
            return nil
        }
        return try? JSONDecoder().decode(T.self, from: octets)
    }

    public func oublier() {
        lock.lock()
        defer { lock.unlock() }
        cachedEntries = nil
        cachedOccurrences = nil
    }

    /// Voir `DiskCorpusRepository.regarder(_:)` — le lexique suit le corpus,
    /// sans quoi l'aperçu montrerait des fiches d'un autre texte.
    public func regarder(_ nouveau: URL) {
        lock.lock()
        defer { lock.unlock() }
        dossier = nouveau
        cachedEntries = nil
        cachedOccurrences = nil
    }
}

/// Les fiches des noms propres, lues du disque quand elles y sont.
///
/// ## Le défaut que ça ferme
///
/// `shemot.json` n'était pas distribué : les fiches restaient celles de
/// l'installation pendant que le texte se corrigeait en minutes. Un lecteur qui
/// touchait un nom en or obtenait la fiche d'il y a trois semaines — et **rien
/// du tout** pour un Shem apparu depuis. Le cas était réel : `gavriel`,
/// `moshe`, `sinai` et `eliyahu` sont nommés dans le corpus sans fiche
/// distribuée.
///
/// Relevé par un audit externe le 8 septembre 2026 — A10.
///
/// ## Pourquoi lire, et non reconstruire
///
/// Reconstruire les fiches ou l'index dans l'app demanderait de réimplémenter
/// l'indexation du pipeline dans **trois liseuses**, avec la certitude qu'elles
/// divergeront. Le pipeline est le seul endroit où la forme se décide ; les
/// liseuses la lisent.
public final class DiskShemotRepository: ShemotRepository, @unchecked Sendable {
    private let dossier: URL
    private let socle: any ShemotRepository
    private let lock = NSLock()
    private var cached: [ShemEntry]?

    public init(
        dossier: URL = CorpusUpdater.dossierParDefaut(),
        socle: any ShemotRepository = BundleShemotRepository()
    ) {
        self.dossier = dossier
        self.socle = socle
    }

    public func entries() throws -> [ShemEntry] {
        lock.lock()
        defer { lock.unlock() }

        if let cached { return cached }
        let entries: [ShemEntry]
        if let file: ONTSchema.ShemotFile = lire("shemot.json") {
            entries = file.entries.map(ShemEntry.init)
        } else {
            entries = try socle.entries()
        }
        cached = entries
        return entries
    }

    private func lire<T: Decodable>(_ nom: String) -> T? {
        guard let octets = try? Data(contentsOf: dossier.appendingPathComponent(nom)) else {
            return nil
        }
        return try? JSONDecoder().decode(T.self, from: octets)
    }

    /// Oublie ce qui est en mémoire, après une mise à jour.
    public func oublier() {
        lock.lock()
        defer { lock.unlock() }
        cached = nil
    }
}

/// L'index de recherche, lu du disque quand il y est.
///
/// Même raison que les fiches : un index figé à l'installation ne trouve pas
/// ce qui a été écrit depuis. Et il est le plus gros des fichiers du corpus —
/// six cents kilo-octets —, donc le plus susceptible d'échouer sur un lien
/// mauvais. Le bundle répond en attendant, comme partout ailleurs.
///
/// **La clé publiée est `recherche`, et le fichier local `search.json`.** Le
/// premier est le nom du nœud dans le manifeste, écrit en français comme le
/// reste ; le second est le nom que le pipeline donne au fichier depuis le
/// début. Les confondre ferait chercher un fichier qui n'existe pas, et le
/// bundle répondrait à sa place — sans que rien ne le dise.
public final class DiskSearchIndex: SearchIndex, @unchecked Sendable {
    private let dossier: URL
    private let socle: any SearchIndex
    private let lock = NSLock()
    private var cached: [SearchRecord]?

    public init(
        dossier: URL = CorpusUpdater.dossierParDefaut(),
        socle: any SearchIndex = BundleSearchIndex()
    ) {
        self.dossier = dossier
        self.socle = socle
    }

    public func records() -> [SearchRecord] {
        lock.lock()
        defer { lock.unlock() }

        if let cached { return cached }
        let records: [SearchRecord]
        if let file: ONTSchema.SearchFile = lire("search.json") {
            records = file.records.map(SearchRecord.init)
        } else {
            records = socle.records()
        }
        cached = records
        return records
    }

    private func lire<T: Decodable>(_ nom: String) -> T? {
        guard let octets = try? Data(contentsOf: dossier.appendingPathComponent(nom)) else {
            return nil
        }
        return try? JSONDecoder().decode(T.self, from: octets)
    }

    public func oublier() {
        lock.lock()
        defer { lock.unlock() }
        cached = nil
    }
}

/// La feuille de prononciation, du disque quand elle y est, du bundle sinon.
///
/// Même montage que le glossaire, pour la même raison : le texte se corrige, et
/// une feuille figée à l'installation expliquerait la prononciation d'hier.
public final class DiskPrononciationRepository: PrononciationRepository, @unchecked Sendable {
    private let dossier: URL
    private let bundle: Foundation.Bundle
    private let lock = NSLock()
    private var cachee: FeuilleDePrononciation??

    public init(
        dossier: URL = CorpusUpdater.dossierParDefaut(),
        bundle: Foundation.Bundle = .main
    ) {
        self.dossier = dossier
        self.bundle = bundle
    }

    public func feuille() -> FeuilleDePrononciation? {
        lock.lock()
        defer { lock.unlock() }

        if let cachee { return cachee }
        let dto: ONTSchema.PrononciationFile? =
            lire("prononciation.json")
            ?? (try? BundleLoader.decode("prononciation", bundle: bundle))
        let feuille = dto.map {
            FeuilleDePrononciation(titre: $0.title, blocs: $0.blocks.map(Block.init))
        }
        cachee = feuille
        return feuille
    }

    private func lire<T: Decodable>(_ nom: String) -> T? {
        guard let octets = try? Data(contentsOf: dossier.appendingPathComponent(nom)) else {
            return nil
        }
        return try? JSONDecoder().decode(T.self, from: octets)
    }

    public func oublier() {
        lock.lock()
        defer { lock.unlock() }
        cachee = nil
    }
}

File: android/ontdata/src/main/kotlin/com/labibleont/ont/data/remote/DiskRepositories.kt
package com.labibleont.ont.data.remote

import android.content.Context
import com.labibleont.ont.data.bundle.AssetCorpusRepository
import com.labibleont.ont.data.bundle.AssetGlossaryRepository
import com.labibleont.ont.data.bundle.AssetLoader
import com.labibleont.ont.data.bundle.AssetSearchIndex
import com.labibleont.ont.data.bundle.AssetShemotRepository
import com.labibleont.ont.data.bundle.versDomaine
import com.labibleont.ont.data.schema.Book as DtoBook
import com.labibleont.ont.data.schema.CorpusFile
import com.labibleont.ont.data.schema.GlossaryFile
import com.labibleont.ont.data.schema.OccurrencesFile
import com.labibleont.ont.data.schema.PrononciationFile
import com.labibleont.ont.data.schema.SearchFile
import com.labibleont.ont.data.schema.ShemotFile
import com.labibleont.ont.kit.corpus.Book
import com.labibleont.ont.kit.corpus.Corpus
import com.labibleont.ont.kit.corpus.ShemEntry
import com.labibleont.ont.kit.glossary.GlossaryEntry
import com.labibleont.ont.kit.glossary.Occurrence
import com.labibleont.ont.kit.ports.FeuilleDePrononciation
import com.labibleont.ont.kit.ports.CorpusRepository
import com.labibleont.ont.kit.ports.GlossaryRepository
import com.labibleont.ont.kit.ports.PrononciationRepository
import com.labibleont.ont.kit.ports.SearchIndex
import com.labibleont.ont.kit.ports.ShemotRepository
import com.labibleont.ont.kit.search.SearchRecord
import java.io.File
import kotlinx.serialization.json.Json

/**
 * Le corpus, lu du disque quand il y est, du bundle sinon.
 *
 * ## Le bundle n'est pas un repli, c'est le socle
 *
 * L'app embarque un corpus complet : elle fonctionne au premier lancement, sans
 * réseau, et fonctionnera toujours. **Le disque ne fait que le recouvrir,
 * fichier par fichier.**
 *
 * Un livre téléchargé recouvre son homologue ; les autres continuent de venir du
 * bundle. Un téléchargement raté ne laisse donc aucun trou — il laisse ce qui
 * était là.
 *
 * ## Le recouvrement est par fichier, pas par lot
 *
 * Recouvrir en bloc — « tout le disque ou tout le bundle » — obligerait à
 * décider quoi faire quand sept livres sur huit sont arrivés. Fichier par
 * fichier, la question ne se pose pas : chaque morceau est le plus récent dont
 * on dispose.
 *
 * ## Et le disque ne gagne pas toujours
 *
 * C'est la leçon qui a coûté le plus cher. Le disque l'emportait sans condition,
 * si bien qu'une build neuve se faisait écraser au premier lancement par le
 * corpus publié — plus ancien tant que le site n'avait pas republié. Une build
 * portant 1 913 noms propres en affichait 217.
 *
 * L'arbitrage vit dans [CorpusUpdater.plusRecentQueLeBundle] : le disque n'est
 * rempli que par un manifeste dont on a **prouvé** qu'il est plus récent. Ici,
 * on lit ce qui s'y trouve sans avoir à se reposer la question.
 */
public class DiskCorpusRepository(
    private val context: Context,
    private val dossier: File = CorpusUpdater.dossierParDefaut(context),
) : CorpusRepository {

    private val bundle = AssetCorpusRepository(context)
    private val json = Json { ignoreUnknownKeys = true }

    override fun corpora(): kotlin.collections.List<Corpus> =
        lire<CorpusFile>("corpus.json")?.corpora?.map { it.versDomaine() } ?: bundle.corpora()

    override fun book(id: String): Book =
        lire<DtoBook>("books/$id.json")?.versDomaine() ?: bundle.book(id)

    private inline fun <reified T> lire(nom: String): T? = runCatching {
        val f = File(dossier, nom)
        if (!f.exists()) null else json.decodeFromString<T>(f.readText())
    }.getOrNull()
}

/** Le lexique, disque par-dessus bundle. */
public class DiskGlossaryRepository(
    private val context: Context,
    private val dossier: File = CorpusUpdater.dossierParDefaut(context),
) : GlossaryRepository {

    private val bundle = AssetGlossaryRepository(context)
    private val json = Json { ignoreUnknownKeys = true }

    override fun entries(): kotlin.collections.List<GlossaryEntry> = runCatching {
        val f = File(dossier, "glossary.json")
        if (!f.exists()) null
        else json.decodeFromString<GlossaryFile>(f.readText()).entries.map { it.versDomaine() }
    }.getOrNull() ?: bundle.entries()

    // ## Le demi-mégaoctet était un bon argument, et la paresse le règle
    //
    // Ce bloc disait : « les occurrences restent au bundle tant qu'aucune
    // version téléchargée n'a été lue ». La phrase se contredisait — rien
    // n'allait jamais en lire une. Le fichier fait 492 Ko et la liste du lexique
    // n'en a pas besoin : c'est vrai, et c'est pour ça qu'on ne le charge qu'à
    // la première fiche ouverte, pas pour ça qu'on renonce au corpus à jour.
    //
    // iOS lit du disque depuis toujours, paresseusement lui aussi. Après une
    // mise à jour du corpus, Android servait donc des occurrences périmées
    // pendant que le lexique, lui, était neuf.
    private val duDisque: Map<String, kotlin.collections.List<Occurrence>>? by lazy {
        runCatching {
            val f = File(dossier, "occurrences.json")
            if (!f.exists()) null
            else json.decodeFromString<OccurrencesFile>(f.readText())
                .byLemma.mapValues { (_, v) -> v.map { it.versDomaine() } }
        }.getOrNull()
    }

    override fun occurrences(lemma: String): kotlin.collections.List<Occurrence> =
        duDisque?.get(lemma) ?: bundle.occurrences(lemma)
}

/**
 * L'index de recherche, disque par-dessus bundle.
 *
 * ## Il n'existait pas, et rien ne le disait
 *
 * `AssetSearchIndex` était branché directement dans `MainActivity`, sans pendant
 * disque — seul port du corpus dans ce cas. La recherche restait donc sur
 * l'index embarqué à la compilation, quelle que soit la fraîcheur du reste.
 *
 * Le défaut se doublait en amont : `CorpusUpdater` ne déclarait pas
 * `"recherche"`, donc le fichier n'était même pas téléchargé. Réparer un seul
 * des deux n'aurait rien donné — d'où les deux dans le même changement.
 *
 * Chargement paresseux : `search.json` fait 640 Ko, et l'ouverture de l'app n'en
 * a pas besoin.
 */
public class DiskSearchIndex(
    private val context: Context,
    private val dossier: File = CorpusUpdater.dossierParDefaut(context),
) : SearchIndex {

    private val bundle = AssetSearchIndex(context)
    private val json = Json { ignoreUnknownKeys = true }

    private val duDisque: kotlin.collections.List<SearchRecord>? by lazy {
        runCatching {
            val f = File(dossier, "search.json")
            if (!f.exists()) null
            else json.decodeFromString<SearchFile>(f.readText())
                .records.map { it.versDomaine() }
        }.getOrNull()
    }

    override fun records(): kotlin.collections.List<SearchRecord> =
        duDisque ?: bundle.records()
}

/** Les fiches des noms propres, disque par-dessus bundle. */
public class DiskShemotRepository(
    private val context: Context,
    private val dossier: File = CorpusUpdater.dossierParDefaut(context),
) : ShemotRepository {

    private val bundle = AssetShemotRepository(context)
    private val json = Json { ignoreUnknownKeys = true }

    private val duDisque: Map<String, ShemEntry>? by lazy {
        runCatching {
            val f = File(dossier, "shemot.json")
            if (!f.exists()) null
            else json.decodeFromString<ShemotFile>(f.readText())
                .entries.associate { it.lemma to it.versDomaine() }
        }.getOrNull()
    }

    override fun fiche(lemma: String): ShemEntry? =
        duDisque?.get(lemma) ?: bundle.fiche(lemma)
}

/**
 * La feuille de prononciation, du disque quand elle y est, du bundle sinon.
 *
 * ## Le défaut que ça ferme — #263
 *
 * Le pipeline émet `dist/prononciation.json`, iOS le lit dans sept fichiers, et
 * **Android ne l'ouvrait nulle part**. Le fichier était donc *exclu de la copie*
 * vers les ressources, avec un commentaire disant que c'était un écart de parité
 * et non une décision. Cette classe est ce qui permet de retirer l'exclusion.
 *
 * ## Pourquoi une seule classe et non deux
 *
 * Les autres ports d'ici ont un dépôt bundle et un dépôt disque qui l'enveloppe.
 * Celui-ci suit iOS, où `DiskPrononciationRepository` fait les deux : la feuille
 * est **un seul document**, sans clé ni index, donc il n'y a rien à composer
 * entre les deux sources. Le disque gagne s'il porte le fichier, le bundle
 * répond sinon.
 *
 * ## L'absence est un état, pas un échec
 *
 * `null` quand ni le disque ni le bundle ne portent la feuille : le pipeline
 * n'écrit rien quand le vault ne la porte pas. C'est à l'écran de dire ce qu'il
 * attend, pas à ce lecteur de fabriquer un titre vide — ce qui ferait croire à
 * une panne.
 *
 * `by lazy` mémorise aussi l'absence, et c'est voulu : sans ça, chaque ouverture
 * de l'onglet retenterait deux lectures de fichier pour rien.
 *
 * ## Ce qui reste ouvert, et qui n'est pas d'ici
 *
 * iOS porte un `oublier()` qui vide le cache après une mise à jour du corpus.
 * **Aucun dépôt disque d'Android n'en a** — `DiskGlossaryRepository`,
 * `DiskShemotRepository` et `DiskSearchIndex` mémorisent tous par `by lazy` sans
 * invalidation. Une feuille corrigée n'apparaît donc qu'au prochain lancement.
 * Ajouter l'invalidation ici seulement rendrait ce port incohérent avec ses
 * quatre voisins ; c'est un chantier à part, et il se décide en une fois.
 */
public class DiskPrononciationRepository(
    private val context: Context,
    private val dossier: File = CorpusUpdater.dossierParDefaut(context),
) : PrononciationRepository {

    private val json = Json { ignoreUnknownKeys = true }

    private val feuille: FeuilleDePrononciation? by lazy {
        val dto = duDisque() ?: duBundle()
        dto?.let { FeuilleDePrononciation(titre = it.title, blocs = it.blocks.map { b -> b.versDomaine() }) }
    }

    private fun duDisque(): PrononciationFile? = runCatching {
        val f = File(dossier, "prononciation.json")
        if (!f.exists()) null else json.decodeFromString<PrononciationFile>(f.readText())
    }.getOrNull()

    private fun duBundle(): PrononciationFile? = runCatching {
        AssetLoader.decode<PrononciationFile>(context, "data/prononciation.json")
    }.getOrNull()

    override fun feuille(): FeuilleDePrononciation? = feuille
}
